### PR TITLE
Add macros and ifdefs to control constexpr and conditional explicit

### PR DESCRIPTION
**Pull request by jbcoe**

### CODE REVIEW

File: propagate_const.h [hunk 1]
```diff
@@ -363,3 +413,6 @@ struct greater_equal<experimental::fundamentals_v2::propagate_const<T>> {
 };
 
 }  // end namespace std
+
+#undef PROPAGATE_CONST_PROPAGATE_CONST_CONSTEXPR
```
Comment (viboes) [inline propagate_const.h]: s/PROPAGATE_CONST_PROPAGATE_CONST_CONSTEXPR/PROPAGATE_CONST_CONSTEXPR/

File: propagate_const.h [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
+*/
+
+#ifndef JBCOE_PROPAGATE_CONST_INCLUDED
+#define JBCOE_PROPAGATE_CONST_INCLUDED
 
 #include <functional>
 #include <type_traits>
 #include <utility>
 
+#define PROPAGATE_CONST_CONSTEXPR constexpr
```
Comment (viboes) [inline propagate_const.h]: This should be conditionally defined, isn't it?

File: propagate_const.h [hunk 3]
```diff
@@ -41,84 +67,107 @@ class propagate_const {
...
   // [propagate_const.ctor], constructors
-  constexpr propagate_const() = default;
+  PROPAGATE_CONST_CONSTEXPR propagate_const() = default;
 
   propagate_const(const propagate_const& p) = delete;
 
-  constexpr propagate_const(propagate_const&& p) = default;
+  PROPAGATE_CONST_CONSTEXPR propagate_const(propagate_const&& p) = default;
 
+#ifdef PROPAGATE_CONST_HAS_EXPRESSION_SFINAE
```
Comment (viboes) [inline propagate_const.h]: IMO, there is a missing constructor, with the EXPLICIT variant

```c++
EXPLICIT constexpr propagate_const(propagate_const<U> const& pu);
```

File: propagate_const.h (hunk 1 above)
Comment (jbcoe) [reply propagate_const.h]: fixed. Thanks.

File: propagate_const.h (hunk 2 above)
Comment (jbcoe) [reply propagate_const.h]: Will fix.

File: propagate_const.h [hunk 4]
```diff
@@ -41,84 +67,107 @@ class propagate_const {
...
+#endif
   // [propagate_const.assignment], assignment
   propagate_const& operator=(const propagate_const& p) = delete;
 
-  constexpr propagate_const& operator=(propagate_const&& p) = default;
+  PROPAGATE_CONST_CONSTEXPR propagate_const& operator=(propagate_const&& p) = default;
 
   template <class U>
-  constexpr propagate_const& operator=(propagate_const<U>&& pu) {
+  PROPAGATE_CONST_CONSTEXPR propagate_const& operator=(propagate_const<U>&& pu) {
```
Comment (viboes) [inline propagate_const.h]: Wondering if this overload shouldn't be constrained when the equivalent construction is implicit.

Comment (jbcoe) [reply propagate_const.h]: I think you are right. 

I'll improve the tests next. Then we can iterate on improving things.
Tests from libcxx are written by me and easy to re-purpose.

File: propagate_const.h (hunk 3 above)
Comment (jbcoe) [reply propagate_const.h]: Copy constructors are deleted for propagate const as it lets one get round const-propagation.

Comment (viboes) [reply propagate_const.h]: yes, I forgot about that.

**Comment by philipcraig** [comment]
If this pull request is meant to make the code viable on all of: VS2015, VS2017, gcc 4.9, gcc6, then where are the .appveyor.yml and .travis.yml changes that enforce this?

File: propagate_const.h [hunk 5]
```diff
@@ -73,7 +80,24 @@ class propagate_const {
 
   PROPAGATE_CONST_CONSTEXPR propagate_const(propagate_const&& p) = default;
 
-#ifdef PROPAGATE_CONST_HAS_EXPRESSION_SFINAE
+#ifdef PROPAGATE_CONST_HAS_NO_EXPRESSION_SFINAE
+  //
+  // Make converting constructors explicit as we cannot use SFINAE to check.
+  //
+  template <class U, class = enable_if_t<is_constructible<T, U&&>::value>>
```
Comment (viboes) [inline propagate_const.h]: Have you tried with 

```
template <class U, enable_if_t<is_constructible<T, U&&>::value, int>=0>
```
?

Comment (viboes) [reply propagate_const.h]: Or is this exactly what MSVC donesn't like?

**Comment by jbcoe** [comment]
@philipcraig They will need merging in next.

File: propagate_const.h (hunk 5 above)
Comment (jbcoe) [reply propagate_const.h]: I’m happy to iterate and improve.

**Pull request merged**